### PR TITLE
Rewards bitflyer android integration

### PR DESCRIPTION
**Pull request by deeppandya**
Resolves https://github.com/brave/brave-browser/issues/18471

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_rewards/browser/rewards_protocol_handler.cc [hunk 1]
```diff
@@ -69,7 +69,7 @@ void LoadRewardsURL(
   const std::string bitflyer_staging_host = bitflyer_staging_url.host();
   const std::string gemini_staging_host = gemini_oauth_staging_url.host();
   const char* kAllowedDomains[] = {
-      "bitflyer.com",                 // bitFlyer production
+      "dummy.com",                    // bitFlyer production
```
Comment (SergeyZhukovsky) [inline components/brave_rewards/browser/rewards_protocol_handler.cc:72]: is it a real domain?

Comment (emerick) [reply components/brave_rewards/browser/rewards_protocol_handler.cc:72]: No, the bitFlyer production domain is stubbed out for now per request from @Miyayes.

**Comment by SergeyZhukovsky** [comment]
@deeppandya I think we need to check it on Android 6, as the PR contains new vector images.

**Comment by deeppandya** [comment]
> @deeppandya I think we need to check it on Android 6, as the PR contains new vector images.

@SergeyZhukovsky I have checked. If we use support components, it's fine.

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/BraveRewardsVerifyWalletActivity.java [hunk 2]
```diff
@@ -63,7 +97,8 @@ void SetUpholdLinkHandler() {
 
                         if (BraveRewardsHelper.subtextAtOffset(built_service_str, part2, offset) ){
                             Intent intent = new Intent();
-                            intent.putExtra(BraveActivity.OPEN_URL, BraveUphold.UPHOLD_ORIGIN_URL);
+                            intent.putExtra(
+                                    BraveActivity.OPEN_URL, BraveWalletProvider.UPHOLD_ORIGIN_URL);
```
Comment (emerick) [inline android/java/org/chromium/chrome/browser/BraveRewardsVerifyWalletActivity.java:101]: @deeppandya If this function is Uphold-specific, do we need a bitFlyer-specific function. If it's not Uphold-specific, then it seems like we need to be smarter here (and probably should rename the function, etc.)

File: android/java/org/chromium/chrome/browser/BraveWalletProvider.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    private String parseJsonArgs(String jsonArgs) {
+        String redirect_url = "";
+        try {
+            JSONObject jsonObj = new JSONObject(jsonArgs);
+            if (jsonObj.has(REDIRECT_URL_KEY)) {
+                redirect_url = jsonObj.getString(REDIRECT_URL_KEY);
+            }
+        } catch (JSONException e) {
+        }
```
Comment (emerick) [inline android/java/org/chromium/chrome/browser/BraveWalletProvider.java:115]: Should we log something here?

File: android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java [hunk 4]
```diff
@@ -0,0 +1,1646 @@
...
+        setVerifyWalletButton(walletStatus);
+    }
+
+    private void setVerifyWalletButton(@WalletStatus final int status) {
+        TextView btnVerifyWallet = mPopupView.findViewById(R.id.btn_verify_wallet);
+        SharedPreferences sharedPref = ContextUtils.getAppSharedPreferences();
+        SharedPreferences.Editor editor = sharedPref.edit();
+        int rightDrawable = 0;
+        int leftDrawable = 0;
+        int text = 0;
```
Comment (emerick) [inline android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1158]: Nit: Maybe `textId` is a better name here?

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 5]
```diff
@@ -170,6 +170,8 @@
     public static final String ADD_FUNDS_URL = "chrome://rewards/#add-funds";
     public static final String REWARDS_SETTINGS_URL = "chrome://rewards/";
     public static final String BRAVE_REWARDS_SETTINGS_URL = "brave://rewards/";
+    public static final String BRAVE_REWARDS_SETTINGS_WALLET_PROVIDER_URL =
```
Comment (emerick) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:173]: Nit: I would name this `BRAVE_REWARDS_SETTINGS_WALLET_VERIFICATION_URL` just to be a little more specific.

File: android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java [hunk 6]
```diff
@@ -0,0 +1,1646 @@
...
+
+                if (android.os.Build.VERSION.SDK_INT
+                        < android.os.Build.VERSION_CODES.N) { // Before 7.0
+                    loginPopupWindow.showAsDropDown(view, 0, 0);
+                } else {
+                    int[] location = new int[2];
+                    view.getLocationOnScreen(location);
+                    int x = location[0];
+                    int y = location[1];
+                    loginPopupWindow.showAtLocation(view, Gravity.NO_GRAVITY, 0, y);
```
Comment (emerick) [inline android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1606]: Just checking that we don't want to use `x` here?

File: android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java [hunk 7]
```diff
@@ -0,0 +1,1646 @@
...
+                double usdValue = walletBalance * mBraveRewardsNativeWorker.GetWalletRate();
+                String usdText =
+                        String.format(mPopupView.getResources().getString(R.string.brave_ui_usd),
+                                String.format(Locale.getDefault(), "%.2f", usdValue));
+                TextView usdBalanceText = mPopupView.findViewById(R.id.usd_balance_text);
+                usdBalanceText.setText(usdText);
+            }
+        } else if (errorCode == BraveRewardsNativeWorker.LEDGER_ERROR) { // No Internet connection
+            String args[] = {};
+            Log.e(TAG, "LEDGER_ERROR");
```
Comment (emerick) [inline android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1109]: Maybe we could say something like `Failed to fetch rewards parameters from server` (or something like that) here?

**Review by zenparsing — APPROVED** [review]

File: components/brave_rewards/browser/rewards_service.h [hunk 8]
```diff
@@ -312,6 +312,8 @@ class RewardsService : public KeyedService {
 
   virtual void GetExternalWallet(GetExternalWalletCallback callback) = 0;
 
+  virtual std::string GetExternalWalletType() const = 0;
```
Comment (zenparsing) [inline components/brave_rewards/browser/rewards_service.h:315]: Code that relies on this method may need to be rewritten when we improve how we handle external wallets internally.

Currently, `SetExternalWalletType` is used to set a pref, which mostly serves to control which wallet type will be returned from `GetExternalWallet`. This was a hack to work around the fact that the front-end design expects to be able to retrieve a populated `mojom::ExternalWallet` structure, even when the user is not connected to any wallet provider.

Ideally, this pref would not be necessary: the engine would know which wallet is currently connected (if any), and the front-end would not expect to get "login" URLs from the `ExternalWallet` structure. In that case we'd remove the `SetExternalWalletType` method.

Code that relies on `GetExternalWalletType` (a sync method), may have to be rewritten in the future, and I would prefer that we not have more implementations using the `SetExternalWalletType` hack going forward.

File: components/brave_rewards/browser/rewards_protocol_handler.cc (hunk 1 above)
Comment (emerick) [reply components/brave_rewards/browser/rewards_protocol_handler.cc]: @zenparsing rightly pointed out that this won't work as-is, so we're reverting it and will rely on stubbing the production client id in our builds instead.

File: components/brave_rewards/browser/rewards_service.h (hunk 8 above)
Comment (emerick) [reply components/brave_rewards/browser/rewards_service.h:315]: @zenparsing mentioned in our meeting that this is good for now, but let's not use `GetExternalWalletType` beyond this going forward.

**Pull request merged**